File: ios/Classes/LeancloudFlutterPlugin.m
#import "LeancloudFlutterPlugin.h"

#import "LeancloudFunction.h"
#import "LeancloudObject.h"
#import "LeancloudQuery.h"

@implementation LeancloudFlutterPlugin

LeancloudFunction *leancloudFunction;

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"leancloud_flutter_plugin"
                                     binaryMessenger:[registrar messenger]];
    LeancloudFlutterPlugin* instance = [[LeancloudFlutterPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
    leancloudFunction = [[LeancloudFunction alloc] init];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    if ([@"initialize" isEqualToString:call.method]) {
        [leancloudFunction initialize:call result:result];
    } else if ([@"setLogLevel" isEqualToString:call.method]) {
        [leancloudFunction setAllLogsEnabled:call result:result];
    } else if ([@"setRegion" isEqualToString:call.method]) {
        // Auto set region
        result(nil);
    } else if ([@"saveOrCreate" isEqualToString:call.method]) {
        LeancloudObject *leancloudObject = [[LeancloudObject alloc] init];
        [leancloudObject saveOrCreate:call result:result];
    } else if ([@"delete" isEqualToString:call.method]) {
        LeancloudObject *leancloudObject = [[LeancloudObject alloc] init];
        [leancloudObject deleteObject:call result:result];
    } else if ([@"query" isEqualToString:call.method]) {
        LeancloudQuery *leancloudQuery = [[LeancloudQuery alloc] init];
        [leancloudQuery query:call result:result];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

@end

File: android/src/main/java/com/lunagao/leancloudflutterplugin/LeancloudFlutterPlugin.java
package com.lunagao.leancloudflutterplugin;

import android.content.Context;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** LeancloudFlutterPlugin */
public class LeancloudFlutterPlugin implements MethodCallHandler {

  private static Context _applicationContext;

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "leancloud_flutter_plugin");
    channel.setMethodCallHandler(new LeancloudFlutterPlugin());
    _applicationContext = registrar.context().getApplicationContext();
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    switch (call.method) {
      case "initialize":
        LeancloudFunction.initialize(call, result, _applicationContext);
        break;
      case "setLogLevel":
        LeancloudFunction.setLogLevel(call, result);
        break;
      case "setRegion":
        LeancloudFunction.setRegion(call, result);
        break;
      case "saveOrCreate":
        LeancloudObject saveOrCreateObject = new LeancloudObject();
        saveOrCreateObject.saveOrCreate(call, result);
        break;
      case "delete":
        LeancloudObject deleteObject = new LeancloudObject();
        deleteObject.delete(call, result);
        break;
      case "query":
        LeancloudQuery leancloudQuery = new LeancloudQuery();
        leancloudQuery.query(call, result);
        break;
      case "doCloudQuery":
        LeancloudQuery doCloudQueryQuery = new LeancloudQuery();
        doCloudQueryQuery.doCloudQuery(call, result);
        break;
      default:
        result.notImplemented();
    }
  }
}
